analysis/scala-implicit-browser/app.R

```diff
@@ -0,0 +1,350 @@
+          choices=NULL
+        ),
+        h4("Filter"),
+        checkboxGroupInput(
+          inputId="filter_kind",
+          label=h5("by kind")
+        ),
+        checkboxGroupInput(
+          inputId="filter_srcset",
+          label=h5("by srcset")
+        ),
+        checkboxGroupInput(
+          inputId="filter_module",
+          label=h5("by module")
+        ),
+        checkboxGroupInput(
+            inputId="filter_param_type",
+          label=h5("by param type")
+        ),
+        hr()
+      ),
+
+      mainPanel(
+        h3("Graph"),
+        visNetworkOutput("graph", height = "600px"),
+        h3("Selected Node"),
+        dataTableOutput("node_details"),
+        h3("Neighbors"),
+        dataTableOutput("node_neighbors")
+      )
+   )
+)
+
+server <- function(input, output, session) {
+  data <- reactive(load_data())
+
+  observe({
+    data <- data()
+    df <- data$projects %>%
+      mutate(name=str_c(name, version, sep=":")) %>%
+      select(name, id)
+
+    projects <- df$id
+    names(projects) <- df$name
+
+    updateSelectInput(
+      session,
+      "project",
+      choices=projects,
+      selected=projects[1]
+    )
+  })
+
+  model <- reactive({
+    data <- data()
+    p_id <- as.integer(input$project)
+
+    if (is.na(p_id)) {
+      NULL
+    } else {
+      project <- data$project %>% filter(id == p_id)
+
+      funs <-
+        data$funs %>%
+        filter(project == p_id) %>%
+        collect(n=Inf)
+
+      params <-
+        data$params %>%
+        filter(project == p_id) %>%
+        collect(n=Inf) %>%
+        distinct(id, .keep_all=TRUE)
+
+      # there is a bug in DPLYR which generates wring query
+      params_funs <-
+        tbl(
+          data$con,
+          sql(str_c("select pf.* from params p inner join params_funs pf on p.id=pf.param where p.project=", p_id))
+        ) %>%
+        collect(n=Inf)
+
+      create_model(project, funs, params, params_funs)
+    }
+  })
+
+  observe({
+    model <- model()
+    kind <- if (is.null(model)) {
+      NULL
+    } else {
+      model$nodes %>%
+        filter(node_type == NODE_PARAM) %>%
+        select(kind) %>%
+        distinct() %>%
+        .$kind
+    }
+
+    updateCheckboxGroupInput(
+      session,
+      "filter_kind",
+      choices=kind,
+      selected=kind
+    )
+  })
+
+  observe({
+    model <- model()
+    modules <- if (is.null(model)) {
+      NULL
+    } else {
+      model$nodes %>%
+        filter(node_type == NODE_FUN) %>%
+        select(module) %>%
+        distinct() %>%
+        .$module
+    }
+
+    updateCheckboxGroupInput(
+      session,
+      "filter_module",
+      choices=modules,
+      selected=modules
+    )
+  })
+
+  observe({
+    model <- model()
+    srcset <- if (is.null(model)) {
+      NULL
+    } else {
+      model$nodes %>%
+        filter(node_type == NODE_FUN) %>%
+        select(srcset) %>%
+        distinct() %>%
+        .$srcset
+    }
+
+    updateCheckboxGroupInput(
```

Comment:
Slight bug, which may totally come from the data in the database being ill-formatted. If that's the case, we don't have to worry.

![wonky srcsets](https://i.imgur.com/622KQmU.png)

Reply:
Good point, I was wondering when it will show up :-)

Reply:
Fixed in f693744